// twilio-run/dist/config/deploy.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfigFromFlags = void 0;
const path_1 = __importDefault(require("path"));
const deploy_1 = require("../commands/deploy");
const utils_1 = require("../commands/utils");
const utils_2 = require("../serverless-api/utils");
const global_1 = require("./global");
const utils_3 = require("./utils");
const mergeFlagsAndConfig_1 = require("./utils/mergeFlagsAndConfig");
const userAgentExtensions_1 = require("./utils/userAgentExtensions");
async function getConfigFromFlags(flags, externalCliOptions) {
    let cwd = flags.cwd ? path_1.default.resolve(flags.cwd) : process.cwd();
    flags.cwd = cwd;
    if (typeof flags.functionsEnv !== 'undefined') {
        (0, utils_1.deprecateFunctionsEnv)();
        if (!flags.environment) {
            flags.environment = flags.functionsEnv;
        }
        delete flags.functionsEnv;
    }
    if (flags.production) {
        flags.environment = '';
    }
    const configFlags = (0, global_1.readSpecializedConfig)(cwd, flags.config, 'deploy', {
        username: flags.username ||
            (externalCliOptions && externalCliOptions.accountSid) ||
            undefined,
        environmentSuffix: flags.environment,
        region: flags.region,
    });
    flags = (0, mergeFlagsAndConfig_1.mergeFlagsAndConfig)(configFlags, flags, deploy_1.cliInfo);
    cwd = flags.cwd || cwd;
    const { localEnv: envFileVars, envPath } = await (0, utils_3.readLocalEnvFile)(flags);
    const { username, password } = await (0, utils_3.getCredentialsFromFlags)(flags, envFileVars, externalCliOptions);
    const env = (0, utils_3.filterEnvVariablesForDeploy)(envFileVars);
    const serviceSid = flags.serviceSid ||
        (await (0, utils_2.getFunctionServiceSid)(cwd, flags.config, 'deploy', flags.username && flags.username.startsWith('AC')
            ? flags.username
            : username.startsWith('AC')
                ? username
                : externalCliOptions === null || externalCliOptions === void 0 ? void 0 : externalCliOptions.accountSid, flags.region));
    const pkgJson = await (0, utils_3.readPackageJsonContent)(flags);
    let serviceName = await (0, utils_3.getServiceNameFromFlags)(flags);
    if (serviceSid === null || serviceSid === void 0 ? void 0 : serviceSid.startsWith('ZS')) {
        serviceName = '';
    }
    else if (!serviceName) {
        throw new Error('Please pass --service-name or add a "name" field to your package.json');
    }
    const { region, edge, runtime } = flags;
    const outputFormat = flags.outputFormat || (externalCliOptions === null || externalCliOptions === void 0 ? void 0 : externalCliOptions.outputFormat);
    return {
        cwd,
        envPath,
        username,
        password,
        env,
        serviceSid,
        pkgJson,
        overrideExistingService: flags.overrideExistingProject,
        force: flags.force,
        serviceName,
        functionsEnv: flags.environment,
        functionsFolderName: flags.functionsFolder,
        assetsFolderName: flags.assetsFolder,
        noAssets: !flags.assets,
        noFunctions: !flags.functions,
        region,
        edge,
        runtime,
        userAgentExtensions: (0, userAgentExtensions_1.getUserAgentExtensions)('deploy', externalCliOptions),
        outputFormat,
    };
}
exports.getConfigFromFlags = getConfigFromFlags;
