// twilio-run/dist/config/list.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfigFromFlags = void 0;
const path_1 = __importDefault(require("path"));
const list_1 = require("../commands/list");
const utils_1 = require("../serverless-api/utils");
const global_1 = require("./global");
const utils_2 = require("./utils");
const mergeFlagsAndConfig_1 = require("./utils/mergeFlagsAndConfig");
const userAgentExtensions_1 = require("./utils/userAgentExtensions");
function trim(str) {
    return str.trim();
}
async function getConfigFromFlags(flags, externalCliOptions) {
    var _a;
    let cwd = flags.cwd ? path_1.default.resolve(flags.cwd) : process.cwd();
    flags.cwd = cwd;
    const configFlags = (0, global_1.readSpecializedConfig)(cwd, flags.config, 'list', {
        username: flags.username ||
            (externalCliOptions && externalCliOptions.accountSid) ||
            undefined,
        environmentSuffix: flags.environment,
        region: flags.region,
    });
    flags = (0, mergeFlagsAndConfig_1.mergeFlagsAndConfig)(configFlags, flags, list_1.cliInfo);
    cwd = flags.cwd || cwd;
    const { localEnv: envFileVars, envPath } = await (0, utils_2.readLocalEnvFile)(flags);
    const { username, password } = await (0, utils_2.getCredentialsFromFlags)(flags, envFileVars, externalCliOptions);
    const serviceSid = flags.serviceSid ||
        (await (0, utils_1.getFunctionServiceSid)(cwd, flags.config, 'list', ((_a = flags.username) === null || _a === void 0 ? void 0 : _a.startsWith('AC'))
            ? flags.username
            : username.startsWith('AC')
                ? username
                : externalCliOptions === null || externalCliOptions === void 0 ? void 0 : externalCliOptions.accountSid, flags.region));
    let serviceName = await (0, utils_2.getServiceNameFromFlags)(flags);
    const types = flags.types.split(',').map(trim);
    const region = flags.region;
    const edge = flags.edge;
    const outputFormat = flags.outputFormat || (externalCliOptions === null || externalCliOptions === void 0 ? void 0 : externalCliOptions.outputFormat);
    return {
        cwd,
        username,
        password,
        serviceSid,
        serviceName,
        environment: flags.environment,
        properties: flags.properties
            ? flags.properties.split(',').map((x) => x.trim())
            : undefined,
        extendedOutput: flags.extendedOutput,
        types,
        region,
        edge,
        outputFormat,
        userAgentExtensions: (0, userAgentExtensions_1.getUserAgentExtensions)('list', externalCliOptions),
    };
}
exports.getConfigFromFlags = getConfigFromFlags;
