// twilio-run/dist/config/logs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfigFromFlags = void 0;
const path_1 = __importDefault(require("path"));
const check_service_sid_1 = __importDefault(require("../checks/check-service-sid"));
const logs_1 = require("../commands/logs");
const utils_1 = require("../commands/utils");
const utils_2 = require("../serverless-api/utils");
const global_1 = require("./global");
const utils_3 = require("./utils");
const mergeFlagsAndConfig_1 = require("./utils/mergeFlagsAndConfig");
const userAgentExtensions_1 = require("./utils/userAgentExtensions");
async function getConfigFromFlags(flags, externalCliOptions) {
    var _a;
    let cwd = flags.cwd ? path_1.default.resolve(flags.cwd) : process.cwd();
    flags.cwd = cwd;
    let environment = flags.environment || 'dev';
    flags.environment = environment;
    const configFlags = (0, global_1.readSpecializedConfig)(cwd, flags.config, 'logs', {
        username: flags.username ||
            (externalCliOptions && externalCliOptions.accountSid) ||
            undefined,
        environmentSuffix: environment,
        region: flags.region,
    });
    flags = (0, mergeFlagsAndConfig_1.mergeFlagsAndConfig)(configFlags, flags, logs_1.cliInfo);
    cwd = flags.cwd || cwd;
    environment = flags.environment || environment;
    if (flags.production) {
        environment = '';
    }
    const { localEnv: envFileVars, envPath } = await (0, utils_3.readLocalEnvFile)(flags);
    const { username, password } = await (0, utils_3.getCredentialsFromFlags)(flags, envFileVars, externalCliOptions);
    const command = (0, utils_1.getFullCommand)(flags);
    const potentialServiceSid = flags.serviceSid ||
        (await (0, utils_2.getFunctionServiceSid)(cwd, flags.config, 'logs', ((_a = flags.username) === null || _a === void 0 ? void 0 : _a.startsWith('AC'))
            ? flags.username
            : username.startsWith('AC')
                ? username
                : externalCliOptions === null || externalCliOptions === void 0 ? void 0 : externalCliOptions.accountSid, flags.region));
    const serviceSid = (0, check_service_sid_1.default)(command, potentialServiceSid);
    const outputFormat = flags.outputFormat || (externalCliOptions === null || externalCliOptions === void 0 ? void 0 : externalCliOptions.outputFormat);
    const region = flags.region;
    const edge = flags.edge;
    return {
        cwd,
        username,
        password,
        environment,
        serviceSid,
        outputFormat,
        filterByFunction: flags.functionSid,
        tail: flags.tail,
        region,
        edge,
        logCacheSize: flags.logCacheSize,
        userAgentExtensions: (0, userAgentExtensions_1.getUserAgentExtensions)('logs', externalCliOptions),
    };
}
exports.getConfigFromFlags = getConfigFromFlags;
