// twilio-run/dist/commands/list.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.builder = exports.describe = exports.command = exports.cliInfo = exports.handler = void 0;
const serverless_api_1 = require("@twilio-labs/serverless-api");
const check_credentials_1 = require("../checks/check-credentials");
const check_service_sid_1 = __importDefault(require("../checks/check-service-sid"));
const legacy_config_1 = __importDefault(require("../checks/legacy-config"));
const list_1 = require("../config/list");
const flags_1 = require("../flags");
const list_2 = require("../printers/list");
const logger_1 = require("../utils/logger");
const utils_1 = require("./utils");
const debug = (0, logger_1.getDebugFunction)('twilio-run:list');
function logError(msg) {
    logger_1.logger.error(msg);
}
function handleError(err) {
    debug('%O', err);
    if (err.name === 'TwilioApiError') {
        (0, logger_1.logApiError)(logger_1.logger, err);
    }
    else {
        logError(err.message);
    }
    process.exit(1);
}
async function handler(flags, externalCliOptions) {
    (0, logger_1.setLogLevelByName)(flags.logLevel);
    await (0, legacy_config_1.default)(flags.cwd, false);
    let config;
    try {
        config = await (0, list_1.getConfigFromFlags)(flags, externalCliOptions);
    }
    catch (err) {
        debug(err);
        logError(err.message);
        process.exit(1);
        return;
    }
    if (!config) {
        logError('Internal Error');
        process.exit(1);
        return;
    }
    (0, check_credentials_1.checkConfigForCredentials)(config);
    if (config.types !== 'services' && config.types[0] !== 'services') {
        const command = (0, utils_1.getFullCommand)(flags);
        (0, check_service_sid_1.default)(command, config.serviceSid);
    }
    try {
        const client = new serverless_api_1.TwilioServerlessApiClient(config);
        const result = await client.list({ ...config });
        (0, list_2.printListResult)(result, config, config.outputFormat);
    }
    catch (err) {
        handleError(err);
    }
}
exports.handler = handler;
exports.cliInfo = {
    argsDefaults: {
        types: 'services',
    },
    options: {
        ...(0, flags_1.getRelevantFlags)([
            ...flags_1.BASE_CLI_FLAG_NAMES,
            ...flags_1.BASE_API_FLAG_NAMES,
            'service-name',
            'properties',
            'extended-output',
            'service-sid',
            'output-format',
        ]),
        environment: {
            ...flags_1.ALL_FLAGS['environment'],
            describe: 'The environment to list variables for',
            default: 'dev',
        },
    },
};
function optionBuilder(yargs) {
    if (exports.cliInfo.argsDefaults) {
        yargs = Object.keys(exports.cliInfo.argsDefaults).reduce((yargs, name) => {
            if (exports.cliInfo.argsDefaults && exports.cliInfo.argsDefaults[name]) {
                return yargs.default(name, exports.cliInfo.argsDefaults[name]);
            }
            return yargs;
        }, yargs);
    }
    yargs = yargs
        .example('$0 list services', 'Lists all existing services/projects associated with your Twilio Account')
        .example('$0 ls functions,assets --environment=dev --service-name=demo', 'Lists all existing functions & assets associated with the `dev` environment of this service/project')
        .example('$0 ls environments --service-sid=ZSxxxxx --extended-output', 'Outputs all environments for a specific service with extended output for better parsing')
        .example('$0 ls assets,variables,functions --properties=sid,date_updated', 'Only lists the SIDs and date of last update for assets, variables and function');
    yargs = Object.keys(exports.cliInfo.options).reduce((yargs, name) => {
        return yargs.option(name, exports.cliInfo.options[name]);
    }, yargs);
    return yargs;
}
exports.command = ['list [types]', 'ls [types]'];
exports.describe = 'List existing services, environments, variables, deployments for your Twilio Serverless Account';
exports.builder = optionBuilder;
