// twilio-run/dist/runtime/internal/route-cache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCachedResources = exports.getRoutes = exports.setRoutes = exports.getRouteMap = void 0;
const runtime_paths_1 = require("./runtime-paths");
const allRoutes = new Map();
const assetsCache = new Set();
const functionsCache = new Set();
async function getRouteMap(config) {
    const searchConfig = {};
    if (config.functionsFolderName) {
        searchConfig.functionsFolderNames = [config.functionsFolderName];
    }
    if (config.assetsFolderName) {
        searchConfig.assetsFolderNames = [config.assetsFolderName];
    }
    let routes = await (0, runtime_paths_1.getFunctionsAndAssets)(config.baseDir, searchConfig);
    return setRoutes(routes);
}
exports.getRouteMap = getRouteMap;
function setRoutes({ functions, assets }) {
    allRoutes.clear();
    assetsCache.clear();
    functionsCache.clear();
    functions.forEach((fn) => {
        if (!fn.path) {
            return;
        }
        if (allRoutes.has(fn.path)) {
            throw new Error(`Duplicate. Path ${fn.path} already exists`);
        }
        functionsCache.add(fn);
        allRoutes.set(fn.path, {
            ...fn,
            type: 'function',
        });
    });
    assets.forEach((asset) => {
        if (!asset.path) {
            return;
        }
        if (allRoutes.has(asset.path)) {
            throw new Error(`Duplicate. Path ${asset.path} already exists`);
        }
        assetsCache.add(asset);
        allRoutes.set(asset.path, {
            ...asset,
            type: 'asset',
        });
    });
    return new Map(allRoutes);
}
exports.setRoutes = setRoutes;
function getRoutes() {
    return new Map(allRoutes);
}
exports.getRoutes = getRoutes;
function getCachedResources() {
    return {
        assets: Array.from(assetsCache),
        functions: Array.from(functionsCache),
    };
}
exports.getCachedResources = getCachedResources;
